lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1827,28 +1859,64 @@ class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
                                  PatternRewriter &rewriter) const {
     // Collect pipeline registers for stage.
     auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
-
-    // Collect group names for the prologue or epilogue.
-    SmallVector<StringAttr> prologueGroups;
-    SmallVector<StringAttr> epilogueGroups;
-
     // Get the number of pipeline stages in the stages block, excluding the
     // terminator. The verifier guarantees there is at least one stage followed
     // by a terminator.
     size_t numStages = whileOp.getStagesBlock().getOperations().size() - 1;
     assert(numStages > 0);
 
-    // For each registered stage result value.
-    for (auto &operand :
-         stage.getBodyBlock().getTerminator()->getOpOperands()) {
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups, epilogueGroups;
+
+    auto updatePrologueAndEpilogue = [&](calyx::GroupOp group) {
+      // Mark the group for scheduling in the pipeline's block.
+      getComponentState().addBlockScheduleable(stage->getBlock(), group);
+
+      // Add the group to the prologue or epilogue for this stage as
+      // necessary. The goal is to fill the pipeline so it will be in steady
+      // state after the prologue, and drain the pipeline from steady state in
+      // the epilogue. Every stage but the last should have its groups in the
+      // prologue, and every stage but the first should have its groups in the
+      // epilogue.
+      unsigned stageNumber = stage.getStageNumber();
+      if (stageNumber < numStages - 1)
+        prologueGroups.push_back(group.sym_nameAttr());
+      if (stageNumber > 0)
+        epilogueGroups.push_back(group.sym_nameAttr());
+    };
+
+    MutableArrayRef<OpOperand> operands =
+        stage.getBodyBlock().getTerminator()->getOpOperands();
+    bool isStageWithNoPipelinedValues =
+        operands.empty() && !stage.getBodyBlock().empty();
+    if (isStageWithNoPipelinedValues) {
+      // Covers the case where there are no values that need to be passed
+      // through to the next stage, e.g., some intermediary store.
+      auto it = llvm::find_if(stage.getBodyBlock(), [&](auto &op) {
+        Optional<calyx::GroupOp> group =
+            getComponentState().getNonPipelinedGroupFrom<calyx::GroupOp>(&op);
+
+        if (!group.hasValue())
+          return false;
+
+        updatePrologueAndEpilogue(*group);
+        return true;
```

Comment:
Does this assume there is only one store per group? I don't remember if this pass is going to guarantee that or not... Can you please add a test case or update your test case to include two stores in the same stage?

Reply:
Good catch! Fixed, updated the test case.